File: FBSimulatorControlTests/Utilities/FBSimulatorSetTestCase.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorSetTestCase.h"

#import <FBSimulatorControl/FBSimulatorControl.h>

#import "CoreSimulatorDoubles.h"
#import "FBSimulatorControlFixtures.h"

@interface FBSimulatorSetTestCase ()

@end

@implementation FBSimulatorSetTestCase

- (NSArray<FBSimulator *> *)createSetWithExistingSimDeviceSpecs:(NSArray<NSDictionary<NSString *, id> *> *)simulatorSpecs
{
  NSMutableArray<SimDevice *> *simDevices = [NSMutableArray array];
  for (NSDictionary<NSString *, id> *simulatorSpec in simulatorSpecs) {
    FBDeviceModel name = simulatorSpec[@"name"];
    NSUUID *uuid = simulatorSpec[@"uuid"] ?: [NSUUID UUID];
    FBOSVersionName os = simulatorSpec[@"os"] ?: FBOSVersionNameiOS_9_0;
    NSString *version = [[os componentsSeparatedByCharactersInSet:NSCharacterSet.whitespaceCharacterSet] lastObject];
    FBiOSTargetState state = [(simulatorSpec[@"state"] ?: @(FBiOSTargetStateShutdown)) unsignedIntegerValue];

    FBSimulatorControlTests_SimDeviceType_Double *deviceType = [FBSimulatorControlTests_SimDeviceType_Double new];
    deviceType.name = name;

    FBSimulatorControlTests_SimDeviceRuntime_Double *runtime = [FBSimulatorControlTests_SimDeviceRuntime_Double new];
    runtime.name = os;
    runtime.versionString = version;

    FBSimulatorControlTests_SimDevice_Double *device = [FBSimulatorControlTests_SimDevice_Double new];
    device.name = name;
    device.UDID = uuid;
    device.state = (unsigned long long) state;
    device.deviceType = deviceType;
    device.runtime = runtime;

    [simDevices addObject:(SimDevice *)device];
  }

  FBSimulatorControlTests_SimDeviceSet_Double *deviceSet = [FBSimulatorControlTests_SimDeviceSet_Double new];
  deviceSet.availableDevices = [simDevices copy];

  FBSimulatorControlConfiguration *configuration = [FBSimulatorControlConfiguration configurationWithDeviceSetPath:nil options:0 logger:nil reporter:nil];
  _set = [FBSimulatorSet setWithConfiguration:configuration deviceSet:(id)deviceSet delegate:nil logger:nil reporter:nil error:nil];

  NSArray<FBSimulator *> *simulators = _set.allSimulators;
  XCTAssertEqual(simulators.count, simDevices.count);

  // Also confirm that the input ordering is the same as the output ordering
  for (NSUInteger index = 0; index < simulators.count; index++) {
    FBDeviceModel expected = simulatorSpecs[index][@"name"];
    FBDeviceModel actual = simulators[index].deviceType.model;
    XCTAssertEqualObjects(expected, actual);
  }

  return simulators;
}

@end

File: FBSimulatorControlTests/Utilities/FBSimulatorSetTestCase.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@testable import FBSimulatorControl
import Foundation
import XCTest

/// A Test Case Template that creates a Set for mocking.
class FBSimulatorSetTestCase: XCTestCase {

  private(set) var set: FBSimulatorSet!

  @discardableResult
  func createSet(withExistingSimDeviceSpecs simulatorSpecs: [[String: Any]]) -> [FBSimulator] {
    var simDevices: [AnyObject] = []
    var availableSimDevices: [AnyObject] = []
    for simulatorSpec in simulatorSpecs {
      let name = simulatorSpec["name"] as! String
      let uuid = simulatorSpec["uuid"] as? NSUUID ?? NSUUID()
      let os = simulatorSpec["os"] as? String ?? "iOS 9.0"
      let version = os.components(separatedBy: CharacterSet.whitespaces).last ?? os
      let hasRuntimeMetadata = simulatorSpec["hasRuntimeMetadata"] as? Bool ?? true
      let isAvailable = simulatorSpec["available"] as? Bool ?? true
      let state: FBiOSTargetState
      if let stateRaw = simulatorSpec["state"] as? UInt {
        state = FBiOSTargetState(rawValue: stateRaw)!
      } else {
        state = .shutdown
      }

      let deviceType = FBSimulatorControlTests_SimDeviceType_Double()
      deviceType.name = name

      let runtime = FBSimulatorControlTests_SimDeviceRuntime_Double()
      runtime.name = os
      runtime.versionString = version

      let device = FBSimulatorControlTests_SimDevice_Double()
      device.name = name
      device.UDID = uuid
      device.state = UInt64(state.rawValue)
      device.deviceType = deviceType
      device.deviceTypeIdentifier = simulatorSpec["deviceTypeIdentifier"] as? String
        ?? "com.apple.CoreSimulator.SimDeviceType.\(name.replacingOccurrences(of: " ", with: "-"))"
      device.runtimeIdentifier = simulatorSpec["runtimeIdentifier"] as? String
        ?? "com.apple.CoreSimulator.SimRuntime.\(os.replacingOccurrences(of: " ", with: "-").replacingOccurrences(of: ".", with: "-"))"
      device.runtime = hasRuntimeMetadata ? runtime : nil

      simDevices.append(device)
      if isAvailable {
        availableSimDevices.append(device)
      }
    }

    let deviceSet = FBSimulatorControlTests_SimDeviceSet_Double()
    deviceSet.availableDevices = availableSimDevices
    deviceSet.devices = simDevices

    let noLogger: (any FBControlCoreLogger)? = nil
    let noReporter: (any FBEventReporter)? = nil
    let configuration = FBSimulatorControlConfiguration(deviceSetPath: nil, logger: noLogger, reporter: noReporter)
    set = CreateSimulatorSetWithFakeDeviceSet(configuration, deviceSet)

    let simulators = set.allSimulators
    XCTAssertEqual(simulators.count, simDevices.count)

    for index in 0..<simulators.count {
      let expected = simulatorSpecs[index]["name"] as! String
      let actual = simulators[index].deviceType.model
      XCTAssertEqual(expected as String, actual.rawValue as String)
    }

    return simulators
  }
}
